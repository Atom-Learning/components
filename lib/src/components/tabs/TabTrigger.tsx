import { Trigger } from '@radix-ui/react-tabs'
import * as React from 'react'
import { styled } from '~/stitches'
import { focusVisibleStyleBlock } from '~/utilities'
import { Text } from '../text'

const StyledTabTrigger = styled(Trigger, {
  background: 'none',
  border: 'none',
  cursor: 'pointer',
  flexShrink: 0,
  p: '$4',
  userSelect: 'none',
  borderBottom: '2px solid transparent',
  '&[data-state="active"]': {
    color: '$interactive1',
    fontWeight: 600,
    letterSpacing: '-0.005em',
    borderColor: 'currentColor'
  },
  '&[data-disabled]': {
    opacity: 0.3,
    cursor: 'not-allowed'
  },
  '&:not([data-disabled])': {
    '&:hover, &:focus-visible': {
      color: '$interactive2',
      bg: '$accentA1'
    },
    '&:active': {
      color: '$interactive3',
    },
    '&:focus-visible': {
      ...focusVisibleStyleBlock(),
    }
  }
})

interface TabTriggerProps
  extends React.ComponentProps<typeof StyledTabTrigger> {
  value: string
}

export const TabTrigger: React.FC<TabTriggerProps> = ({
  children,
  ...rest
}) => (
  <StyledTabTrigger asChild {...rest}><Text size="sm" as="span">{children}</Text></StyledTabTrigger>
)

TabTrigger.displayName = 'TabTrigger'
