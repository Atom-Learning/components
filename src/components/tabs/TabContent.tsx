import * as React from 'react'
import { Content } from '@radix-ui/react-tabs'
import { styled } from '~/stitches'

const StyledTabContent = styled(Content, {
  flexGrow: 1,
  padding: 20
})

type TabContentProps = React.ComponentProps<typeof StyledTabContent> & {
  value: string
}

export const TabContent: React.FC<TabContentProps> = ({
  children,
  css,
  value
}: TabContentProps) => (
  <StyledTabContent value={value} css={css}>
    {children}
  </StyledTabContent>
)

TabContent.displayName = 'TabContent'
