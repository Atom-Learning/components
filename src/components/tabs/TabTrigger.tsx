import { Trigger } from '@radix-ui/react-tabs'
import * as React from 'react'
import { opacify } from 'polished'

import { styled, theme } from '~/stitches'

const StyledTabTrigger = styled(Trigger, {
  textTransform: 'uppercase',
  cursor: 'pointer',
  flexShrink: 0,
  fontFamily: '$body',
  p: '$4',
  height: '$5',
  userSelect: 'none',
  transition: '0.3s',
  variants: {
    theme: {
      light: {
        bg: 'white',
        '&[data-state="active"]': {
          color: '$primary',
          fontWeight: 600,
          letterSpacing: '-0.005em',
          boxShadow: 'inset 0 -2px 0 0 currentColor'
        },
        '&[data-state="inactive"]': {
          color: '$tonal500'
        },
        '&:not([data-disabled]):hover': {
          color: '$primary',
          bg: opacify(-0.9, theme.colors.primary.value)
        },
        '&:not([data-disabled]):active': {
          color: '$primary',
          bg: opacify(-0.8, theme.colors.primary.value),
          boxShadow: 'none'
        },
        '&[data-disabled],&[data-disabled]:hover': {
          color: '$tonal200',
          cursor: 'not-allowed'
        }
      },
      dark: {
        color: 'white',
        '&[data-state="active"]': {
          fontWeight: 600,
          letterSpacing: '-0.005em',
          boxShadow: 'inset 0 -2px 0 0 currentColor'
        },
        '&:not([data-disabled]):hover': {
          bg: opacify(-0.8, 'white')
        },
        '&:not([data-disabled]):active': {
          bg: opacify(-0.7, 'white'),
          boxShadow: 'none'
        },
        '&[data-disabled],&[data-disabled]:hover': {
          color: '$tonal200',
          cursor: 'not-allowed'
        }
      }
    }
  }
})

interface TabTriggerProps
  extends React.ComponentProps<typeof StyledTabTrigger> {
  value: string
  disabled?: boolean
}

export const TabTrigger: React.FC<TabTriggerProps> = ({
  children,
  theme,
  disabled = false,
  ...otherProps
}) => (
  <StyledTabTrigger disabled={disabled} theme={theme} {...otherProps}>
    {children}
  </StyledTabTrigger>
)

TabTrigger.displayName = 'TabTrigger'
