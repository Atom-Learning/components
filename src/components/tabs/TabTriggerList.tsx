import * as React from 'react'
import { List } from '@radix-ui/react-tabs'
import { CSS, styled } from '~/stitches'

const StyledTriggerList = styled(List, {
  flexShrink: 0,
  display: 'flex'
})

type TabTriggerListProps = React.ComponentProps<typeof StyledTriggerList>

export const TabTriggerList: React.FC<TabTriggerListProps> = ({
  children,
  css
}: TabTriggerListProps) => (
  <StyledTriggerList css={css}>{children}</StyledTriggerList>
)

TabTriggerList.displayName = 'TabTriggerList'
