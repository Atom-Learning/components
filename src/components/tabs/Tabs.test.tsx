import { act, fireEvent, render, screen, waitFor } from '@testing-library/react'
import { axe } from 'jest-axe'
import * as React from 'react'
import { IdProvider } from '@radix-ui/react-id'

import { Tabs } from '.'
import { TabContent } from './TabContent'
import { TabTrigger } from './TabTrigger'
import { TabTriggerList } from './TabTriggerList'

const TabsTest = ({ defaultValue = 'tab1' }) => (
  <IdProvider>
    <Tabs defaultValue={defaultValue}>
      <TabTriggerList>
        <TabTrigger value="tab1">Trigger 1</TabTrigger>
        <TabTrigger value="tab2">Trigger 2</TabTrigger>
      </TabTriggerList>
      <TabContent value="tab1">Important content for tab 1</TabContent>
      <TabContent value="tab2">Important content for tab 2</TabContent>
    </Tabs>
  </IdProvider>
)

describe('Tabs component', () => {
  it('renders', async () => {
    const { container } = render(<TabsTest />)

    expect(container).toMatchSnapshot()
  })

  it('has no programmatically detectable a11y issues', async () => {
    const { container } = render(<TabsTest />)

    expect(await axe(container)).toHaveNoViolations()
  })

  it('it makes the correct tab content visible, based on the `defaultValue` prop', async () => {
    render(<TabsTest defaultValue="tab2" />)
    const tab1content = await screen.findByText('Important content for tab 1')
    const tab2content = await screen.findByText('Important content for tab 2')
    expect(tab1content).not.toBeVisible()
    expect(tab2content).toBeVisible()
  })

  it.only('allows the user to switch tabs', async () => {
    render(<TabsTest />)

    const tab1content = await screen.findByText('Important content for tab 1')
    const tab2content = await screen.findByText('Important content for tab 2')
    expect(tab1content).toBeVisible()
    fireEvent.click(screen.queryByText('Trigger 2'))
    screen.debug
    await waitFor(() =>
      expect(screen.queryByText('Important content for tab 2')).toBeVisible()
    )
  })
})
