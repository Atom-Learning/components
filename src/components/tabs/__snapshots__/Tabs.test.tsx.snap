// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component renders 1`] = `
@media  {
  .c-fkcdpq {
    display: flex;
    flex-direction: column;
    padding-top: var(--space-4);
    padding-left: var(--space-2);
    padding-right: var(--space-2);
  }

  .c-bVBcpu {
    flex-shrink: 0;
    display: flex;
  }

  .c-ehnQBI {
    text-transform: uppercase;
    cursor: pointer;
    flex-shrink: 0;
    font-family: var(--fonts-body);
    padding: var(--space-4);
    -webkit-user-select: none;
    user-select: none;
    transition: 0.3s;
  }

  .c-eGaVeY {
    flex-grow: 1;
  }
}

@media  {
  .c-fkcdpq-clEueN-theme-light div[role="tab"] {
    background: white;
  }

  .c-fkcdpq-clEueN-theme-light div[role="tab"][data-state="active"] {
    color: var(--colors-primary);
    text-shadow: 0px 0px 1px currentColor;
    box-shadow: inset 0 -2px 0 0 currentColor;
  }

  .c-fkcdpq-clEueN-theme-light div[role="tab"][data-state="inactive"] {
    color: var(--colors-tonal500);
  }

  .c-fkcdpq-clEueN-theme-light div[role="tab"]:hover {
    color: var(--colors-primary);
    background: rgba(30,113,246,0.1);
  }

  .c-fkcdpq-clEueN-theme-light div[role="tab"]:active {
    color: var(--colors-primary);
    background: rgba(30,113,246,0.2);
    box-shadow: none;
  }

  .c-fkcdpq-clEueN-theme-light div[role="tab"]:focus {
    color: var(--colors-primary);
    box-shadow: inset 0 0 0 2px currentColor;
  }

  .c-fkcdpq-clEueN-theme-light div[role="tab"][data-disabled],
  .c-fkcdpq-clEueN-theme-light div[role="tab"][data-disabled]:hover {
    color: var(--colors-tonal200);
    cursor: default;
  }

  .c-fkcdpq-clEueN-theme-light div[role="tabpanel"] {
    background: white;
    color: var(--colors-textForeground);
    font-family: var(--fonts-body);
  }

  .c-fkcdpq-clEueN-theme-light div[role="tablist"] {
    border-bottom: 1px solid var(--colors-primaryDark);
  }

  .c-fkcdpq-clEueN-theme-light button[role="scrollbar"] {
    background: white;
    opacity: 0.8;
  }
}

<div>
  <div
    class="c-fkcdpq c-fkcdpq-clEueN-theme-light"
    data-orientation="horizontal"
  >
    <div
      aria-orientation="horizontal"
      class="c-bVBcpu"
      data-orientation="horizontal"
      dir="ltr"
      role="tablist"
      style="outline: none;"
      tabindex="0"
    >
      <div
        aria-controls="radix-id-0-1-content-tab1"
        aria-selected="true"
        class="c-ehnQBI"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="active"
        id="radix-id-0-1-trigger-tab1"
        role="tab"
        tabindex="-1"
      >
        Trigger 1
      </div>
      <div
        aria-controls="radix-id-0-1-content-tab2"
        aria-selected="false"
        class="c-ehnQBI"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="inactive"
        id="radix-id-0-1-trigger-tab2"
        role="tab"
        tabindex="-1"
      >
        Trigger 2
      </div>
    </div>
    <div
      aria-labelledby="radix-id-0-1-trigger-tab1"
      class="c-eGaVeY"
      data-orientation="horizontal"
      data-state="active"
      id="radix-id-0-1-content-tab1"
      role="tabpanel"
      tabindex="0"
    >
      Important content for tab 1
    </div>
    <div
      aria-labelledby="radix-id-0-1-trigger-tab2"
      class="c-eGaVeY"
      data-orientation="horizontal"
      data-state="inactive"
      hidden=""
      id="radix-id-0-1-content-tab2"
      role="tabpanel"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`Tabs component renders mobile view 1`] = `
@media  {
  .c-fkcdpq {
    display: flex;
    flex-direction: column;
    padding-top: var(--space-4);
    padding-left: var(--space-2);
    padding-right: var(--space-2);
  }

  .c-bVBcpu {
    flex-shrink: 0;
    display: flex;
  }

  .c-ehnQBI {
    text-transform: uppercase;
    cursor: pointer;
    flex-shrink: 0;
    font-family: var(--fonts-body);
    padding: var(--space-4);
    -webkit-user-select: none;
    user-select: none;
    transition: 0.3s;
  }

  .c-eGaVeY {
    flex-grow: 1;
  }

  .c-dhzjXW {
    display: flex;
  }
}

@media  {
  .c-fkcdpq-clEueN-theme-light div[role="tab"] {
    background: white;
  }

  .c-fkcdpq-clEueN-theme-light div[role="tab"][data-state="active"] {
    color: var(--colors-primary);
    text-shadow: 0px 0px 1px currentColor;
    box-shadow: inset 0 -2px 0 0 currentColor;
  }

  .c-fkcdpq-clEueN-theme-light div[role="tab"][data-state="inactive"] {
    color: var(--colors-tonal500);
  }

  .c-fkcdpq-clEueN-theme-light div[role="tab"]:hover {
    color: var(--colors-primary);
    background: rgba(30,113,246,0.1);
  }

  .c-fkcdpq-clEueN-theme-light div[role="tab"]:active {
    color: var(--colors-primary);
    background: rgba(30,113,246,0.2);
    box-shadow: none;
  }

  .c-fkcdpq-clEueN-theme-light div[role="tab"]:focus {
    color: var(--colors-primary);
    box-shadow: inset 0 0 0 2px currentColor;
  }

  .c-fkcdpq-clEueN-theme-light div[role="tab"][data-disabled],
  .c-fkcdpq-clEueN-theme-light div[role="tab"][data-disabled]:hover {
    color: var(--colors-tonal200);
    cursor: default;
  }

  .c-fkcdpq-clEueN-theme-light div[role="tabpanel"] {
    background: white;
    color: var(--colors-textForeground);
    font-family: var(--fonts-body);
  }

  .c-fkcdpq-clEueN-theme-light div[role="tablist"] {
    border-bottom: 1px solid var(--colors-primaryDark);
  }

  .c-fkcdpq-clEueN-theme-light button[role="scrollbar"] {
    background: white;
    opacity: 0.8;
  }
}

@media  {
  .c-dhzjXW-icmpvrW-css {
    position: relative;
  }

  .c-bVBcpu-ifwuxZe-css {
    width: 100%;
    overflow-x: auto;
  }

  .c-bVBcpu-ifwuxZe-css::-webkit-scrollbar {
    display: none;
  }
}

<div>
  <div
    class="c-fkcdpq c-fkcdpq-clEueN-theme-light"
    data-orientation="horizontal"
  >
    <div
      class="c-dhzjXW c-dhzjXW-icmpvrW-css"
    >
      <div
        aria-orientation="horizontal"
        class="c-bVBcpu c-bVBcpu-ifwuxZe-css"
        data-orientation="horizontal"
        dir="ltr"
        role="tablist"
        style="outline: none;"
        tabindex="0"
      >
        <div
          aria-controls="radix-id-0-1-content-tab1"
          aria-selected="true"
          class="c-ehnQBI"
          data-orientation="horizontal"
          data-radix-collection-item=""
          data-state="active"
          id="radix-id-0-1-trigger-tab1"
          role="tab"
          tabindex="-1"
        >
          Trigger 1 which is going to be long
        </div>
        <div
          aria-controls="radix-id-0-1-content-tab2"
          aria-selected="false"
          class="c-ehnQBI"
          data-orientation="horizontal"
          data-radix-collection-item=""
          data-state="inactive"
          id="radix-id-0-1-trigger-tab2"
          role="tab"
          tabindex="-1"
        >
          Trigger 2 which is going to be even longer
        </div>
      </div>
    </div>
    <div
      aria-labelledby="radix-id-0-1-trigger-tab1"
      class="c-eGaVeY"
      data-orientation="horizontal"
      data-state="active"
      id="radix-id-0-1-content-tab1"
      role="tabpanel"
      tabindex="0"
    >
      Important content for tab 1
    </div>
    <div
      aria-labelledby="radix-id-0-1-trigger-tab2"
      class="c-eGaVeY"
      data-orientation="horizontal"
      data-state="inactive"
      hidden=""
      id="radix-id-0-1-content-tab2"
      role="tabpanel"
      tabindex="0"
    />
  </div>
</div>
`;
